### ISSUE TITLE
EOF error returned by `*ConnPool.Query` 

### ISSUE THREAD

**Issue by fmstephe** [jackc/pgx#337]
It is hard to interpret an io.EOF returned from this method call. Typically an EOF error indicates an end of available input. I wouldn't expect it from such a high level method call.

The question of interpreting this error is made more complex because it does not happen deterministically. It tends to happen only under load and may represent a timeout or the exhaustion of some other resource.

Using delve I have tracked the error being generated here.

```
0  0x000000000056668d in net.(*netFD).Read
at /usr/lib/go/src/net/fd_unix.go:258
1  0x0000000000579c40 in net.(*conn).Read
at /usr/lib/go/src/net/net.go:181
2  0x00000000004631c9 in io.ReadAtLeast
at /usr/lib/go/src/io/io.go:307
3  0x0000000000740ff1 in github.com/jackc/pgx/chunkreader.(*ChunkReader).appendAtLeast
at /home/kpi_service/next/releases/20170929113639/.go/src/github.com/jackc/pgx/chunkreader/chunkreader.go:73
4  0x0000000000740e0d in github.com/jackc/pgx/chunkreader.(*ChunkReader).Next
at /home/kpi_service/next/releases/20170929113639/.go/src/github.com/jackc/pgx/chunkreader/chunkreader.go:63
5  0x00000000007468fb in github.com/jackc/pgx/pgproto3.(*Frontend).Receive
at /home/kpi_service/next/releases/20170929113639/.go/src/github.com/jackc/pgx/pgproto3/frontend.go:50
6  0x00000000007eb361 in github.com/jackc/pgx.(*Conn).rxMsg
at /home/kpi_service/next/releases/20170929113639/.go/src/github.com/jackc/pgx/conn.go:1109
7  0x00000000007e813d in github.com/jackc/pgx.(*Conn).prepareEx
at /home/kpi_service/next/releases/20170929113639/.go/src/github.com/jackc/pgx/conn.go:803
8  0x00000000007fc08a in github.com/jackc/pgx.(*Conn).QueryEx
at /home/kpi_service/next/releases/20170929113639/.go/src/github.com/jackc/pgx/query.go:449
9  0x00000000007fbb1e in github.com/jackc/pgx.(*Conn).Query
at /home/kpi_service/next/releases/20170929113639/.go/src/github.com/jackc/pgx/query.go:329
10  0x00000000007e56a1 in github.com/jackc/pgx.(*Conn).initConnInfo
at /home/kpi_service/next/releases/20170929113639/.go/src/github.com/jackc/pgx/conn.go:393
```

**Comment by jackc** [comment]
Hmm. EOF can occur any place that needs to read the underlying connection if that connection is closed or killed. This is happening while initializing a new connection. It'd be interesting to know if it always occurs here or if it can happen at any time. If only here that makes me think the connection is being canceled by the server in an unexpected way.

Do you have access to the PostgreSQL server logs? I'd be curious if it reported anything interesting.

Another possible option is to pass in a logger to pgx with debug or trace log level. However, these can produce so much log output that it may not be feasible to do in production. It **will** have an impact on performance.

**Comment by fmstephe** [comment]
Some more information.

We are proxying connections through pgbouncer. The pgbouncer logs don't look too unusual. We are getting

```
2017-09-29 17:00:49.649 4708 LOG C-0x1366ce0: (nodb)/(nouser)@127.0.0.1:59576 closing because: client unexpected eof (age=0)
2017-09-29 17:01:19.436 4708 LOG C-0x1366ce0: (nodb)/(nouser)@127.0.0.1:59578 closing because: client unexpected eof (age=0)
2017-09-29 17:01:49.906 4708 LOG C-0x1366ce0: (nodb)/(nouser)@127.0.0.1:59582 closing because: client unexpected eof (age=0)
2017-09-29 17:02:19.866 4708 LOG C-0x1366ce0: (nodb)/(nouser)@127.0.0.1:59584 closing because: client unexpected eof (age=0)
```

Which looks suspicious, but those `eof` logs happen at a steady rate and don't seem to be correlated with the errors are are seeing. I will double check this though, in the past few hours I've been able to refine the reproducing of this error, so I will revisit the pgbouncer logs  next week.

Next week I'll try to trace the queries back to some postgres instances and look at the logs at that end. But I would expect any dropped connections would be logged by pgbouncer.

That stack trace was my first successful identification of an EOF error at source. I'll see if I can detect the errors coming from any other paths.

We are reproducing this on a benchmarking system, so generating huge amounts of logs is not a problem. I'll look into that too.

**Comment by fmstephe** [comment]
A new update.

After a lot of investigation we have found the cause of these errors.

To start we need to say that the stack trace printed with the opening of this issue is misleading. It represents a timeout, but is an artifact of using a debugger and not debugging fast enough. It makes it look as though the timeout occurs while attempting to prepare the SQL query. The EOF errors were generated by the call to `(*ConnPool).Acquire()`. We were able to measure this by managing the connections manually.

The timeout came from `pgbouncer`, which has a 120 second default `query_wait_timeout`, any client query which is not connected to a server within 120 seconds is closed.

In pgbouncer we are using `session` pooling mode. From the pgbouncer documentation mode is described as `Server is released back to pool after client disconnects.` This means that a client connection that is connected to a server connection by pgbouncer will stay connected until the client connection is closed. Unless there are any errors we don't close any client connections.

Our pgbouncer config allows for 50 server connections per server, and our client had 180 connections to the database that was experiencing timeouts.

This isn't a very surprising result, we are using persistent `client->server` connections in pgbouncer and oversubscribing from our client. The first 50 connections will happily connect to the database can happily submit queries. The remaining client connections will never connect to a server and will be closed after the 120 limit.

More interesting is why this behaviour appeared when we switched to `pgx`. Previously we were using the `pq` driver which goes through the stdlib's `database/sql` package. `database/sql` manages database connections for us and does so slightly differently than `pgx`. `database/sql` provides a 'maxIdle` config setting which means that connections are not returned to the connection pool if there are already `maxIdle` connections available. If the connection is not returned to the pool is is closed and discarded. Because our connection pool size was large, `180`, but our `maxIdle` was low, `45`, this meant that we had a steady stream of connections which were closed immediately after use. This caused pgbouncer to connect a waiting client connection to a free server connection.

It is very straightforward to add a `maxIdle` config option to pgx. But in our case the better course of action is to size the client connections to exactly match the server connections in pgbouncer.
